### PR TITLE
KAFKA-2424: Introduce Scalafix linter

### PR DESCRIPTION
**Pull request by jlprat**
Introduces Scalafix, a linter with some rewrite capabilities on
top (https://scalacenter.github.io/scalafix/)

By running the `checkScalafix` gradle tasks a report is printed with all
broken rules.

Running `scalafix` gradle task will apply some rewrites on the files,
for example: import ordering, and annotating return types.
This change uses the official gradle plugin for Scalafix, and Scalafix
itself is maintained by the Scala Center.

Current rules checked:
- Do not allow `final var`
- Explicit return types for public and protected methods (for public
  `val` and `var` the change was too big. This can be done at a later
time)
- Avoid procedure syntax in Scala
- Avoid use of val in for comprehensions (it's a deprecated feature)

*Known limitations*: when automatically inferring the return types,
Scalafix might fail and pick a "too wide" type, i.e. `Any` instead of
the correct one. After letting Scalafix fix the files, contributors
should still look at the changes done and validate they are correct.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by jlprat** [comment]
@cadonna, @ijuma and @vvcephei if any of you has time to take a look at this PR I'd be thankful.

File: Jenkinsfile [hunk 1]
```diff
@@ -20,7 +20,7 @@
 def doValidation() {
   sh """
     ./gradlew -PscalaVersion=$SCALA_VERSION clean compileJava compileScala compileTestJava compileTestScala \
-        spotlessScalaCheck checkstyleMain checkstyleTest spotbugsMain rat \
+        spotlessScalaCheck checkstyleMain checkstyleTest spotbugsMain checkScalafix rat \
```
Comment (jlprat) [inline Jenkinsfile:23]: Added the new task in the Jenkins job script

File: core/src/main/scala/kafka/server/KafkaConfig.scala [hunk 2]
```diff
@@ -1737,11 +1737,11 @@ class KafkaConfig(val props: java.util.Map[_, _], doLog: Boolean, dynamicConfigO
...
 
-  def interBrokerListenerName = getInterBrokerListenerNameAndSecurityProtocol._1
-  def interBrokerSecurityProtocol = getInterBrokerListenerNameAndSecurityProtocol._2
-  def controlPlaneListenerName = getControlPlaneListenerNameAndSecurityProtocol.map { case (listenerName, _) => listenerName }
-  def controlPlaneSecurityProtocol = getControlPlaneListenerNameAndSecurityProtocol.map { case (_, securityProtocol) => securityProtocol }
-  def saslMechanismInterBrokerProtocol = getString(KafkaConfig.SaslMechanismInterBrokerProtocolProp)
+  def interBrokerListenerName: ListenerName = getInterBrokerListenerNameAndSecurityProtocol._1
+  def interBrokerSecurityProtocol: SecurityProtocol = getInterBrokerListenerNameAndSecurityProtocol._2
+  def controlPlaneListenerName: Option[ListenerName] = getControlPlaneListenerNameAndSecurityProtocol.map { case (listenerName, _) => listenerName }
+  def controlPlaneSecurityProtocol: Option[SecurityProtocol] = getControlPlaneListenerNameAndSecurityProtocol.map { case (_, securityProtocol) => securityProtocol }
```
Comment (jlprat) [inline core/src/main/scala/kafka/server/KafkaConfig.scala:1739]: These 2 were incorrectly inferred by Scalafix, I needed to correct them manually.
Scalafix inferred for both `Option[Any]`.

**Comment by jlprat** [comment]
As far as I understand, my jenkinsfile modification will not be taken into consideration by Jenkins in this run, maybe it's worth keep a close eye to this once it's merged

**Comment by jlprat** [comment]
Failure was https://issues.apache.org/jira/browse/KAFKA-13128

**Comment by jlprat** [comment]
Rebased the changes

**Comment by stephan-mohn** [comment]
Aiven Inc - https://aiven.io/ is misusing of Apache Kafka trademark as 'Aiven Kafka' [1] and violates Apache trademarks. See - [1] https://help.aiven.io/en/?q=%22aiven+kafka%22
Aiven Inc continues to misusing Apache trademarks while given multiple warnings in the past from Apache Kafka PMC.
 
CC @saaros, @ivanyu, Apache Legal
 
------------------------------------------------------------------------------------------------------
From: Vxxxx Hasxxxxxx <vahxx@apache.org>
Date: Thu, Aug 26, 2021 at x:xx AM
Subject: Re: Re: Misuse of Apache trademark in Aiven Kafka
To: <viexx@aiven.io>
Cc: <pxxxxxx@kafka.apache.org>, <press@aiven.io>, <emxx@support.aiven.io>


Hi ViXX,

I'm resending my June email to check if there is any progress on your side
updating misuses of Apache Kafka in Aiven communications.
It seems that recent publications continue to use `Aiven Kafka` violating
the use of Apache trademarks: https://help.aiven.io/en/?q=%22aiven+kafka%22

Regards,

--Vxxxx Haxxxxxxx
on behalf of Apache Kafka Project Management Committee

**Pull request closed without merging**